File: frontend-evoltis/src/app/services/clients.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from '../../environments/environment.development';
import { ClientRequest } from '../models/clients/client.request';
import { Observable } from 'rxjs';
import { ClientDto } from '../models/clients/client.dto';
import { format, formatISO } from 'date-fns';
import { ClientsDto } from '../models/clients/clients.dto';

@Injectable({
  providedIn: 'root'
})
export class ClientsService {

  constructor(private http: HttpClient) { }

  postClient(client: ClientRequest): Observable<ClientDto> {
    return this.http.post<ClientDto>(environment.apiUrl + 'clients', client);
  }
  modifyClient(id: string, client: ClientRequest): Observable<ClientDto> {
    return this.http.put<ClientDto>(environment.apiUrl + `clients/${id}`, client);
  }
  getClientById(id: string): Observable<ClientDto> {
    return this.http.get<ClientDto>(environment.apiUrl + `clients/${id}`);
  }

  getClients(aprox: string | null, dateCreatedFrom: Date | null, dateCreatedUntil: Date | null, services: string[] | null): Observable<ClientsDto> {
    let params = new URLSearchParams();
    if(aprox) params.set('Aprox', aprox);
    if(dateCreatedFrom) params.set('DateCreatedFrom', formatISO(dateCreatedFrom, { representation: 'complete' }));
    if(dateCreatedUntil) params.set('DateCreatedUntil', formatISO(dateCreatedUntil, { representation: 'complete' }));
    services?.forEach(service => {
      params.append('Services', service);
  });

    return this.http.get<ClientsDto>(environment.apiUrl + `clients?` + params.toString());
  }

  deleteClient(id: string): Observable<ClientDto> {
    return this.http.delete<ClientDto>(environment.apiUrl + `clients/${id}`);
  }
}

File: backend-evoltis/backend-evoltis.CORE/Services/Imp/ClientsService.cs
using AutoMapper;
using backend_evoltis.CORE.Models.Clients.Requests;
using backend_evoltis.DOMAIN.Entities;
using backend_evoltis.DOMAIN.Interfaces;

namespace backend_evoltis.CORE.Services.Imp
{
    public class ClientsService : IClientService
    {
        private readonly IClientRepository _repository;
        private readonly IServiceRepository _serviceRepository;
        private readonly IMapper _mapper;

        public ClientsService(IClientRepository repository, IMapper mapper, IServiceRepository serviceRepository)
        {
            _repository = repository;
            _mapper = mapper;
            _serviceRepository = serviceRepository;
        }

        public async Task<Client> DeleteClient(Guid id) => await _repository.DeleteClient(id);

        public async Task<Client> GetClientById(Guid id) => await _repository.GetClientById(id);

        public async Task<List<Client>> GetClients(
            string? aprox, 
            DateTime? dateCreatedFrom, 
            DateTime? dateCreatedUntil,
            List<Guid>? services
            ) => await _repository.GetClients(
                aprox, 
                dateCreatedFrom, 
                dateCreatedUntil,
                services
                );

        public async Task<Client> ModifyClient(Guid id, ClientRequest request)
        {
            var client = await _repository.GetClientById(id);
            client = _mapper.Map(request, client);
            client.Services = [];
            foreach (var serviceId in request.Services)
            {
                var service = await _serviceRepository.GetServiceById(serviceId);
                if (client.Services.Any(s => s.Service.Id == serviceId)) continue;
                var clientService = new ClientService
                {
                    Service = service,
                    Client = client
                };
                client.Services.Add(clientService);
            }
            client.ModifiedAt = DateTime.Now;
            return await _repository.ModifyClient(client);
        }

        public async Task<Client> PostClient(ClientRequest request)
        {
            var client = _mapper.Map<Client>(request);
            client.Services = [];
            foreach (var serviceId in request.Services)
            {
                var service = await _serviceRepository.GetServiceById(serviceId);
                var clientService = new ClientService
                {
                    Service = service,
                    Client = client
                };
                client.Services.Add(clientService);
            }
            client.ModifiedAt = DateTime.Now;
            return await _repository.PostClient(client);
        }
    }
}
